vchan/src/lib.rs

```diff
@@ -0,0 +1,208 @@
+/*
+ * The Qubes OS Project, https://www.qubes-os.org
+ *
+ * Copyright (C) 2010  Rafal Wojtczuk  <rafal@invisiblethingslab.com>
+ * Copyright (C) 2021  Demi Marie Obenour  <demi@invisiblethingslab.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ */
+use std::io::{Error, Read, Write};
+use std::os::{raw::c_int, unix::prelude::RawFd};
+
+/// Status of the channel
+#[derive(Debug, PartialEq, Eq, Copy, Clone)]
+pub enum Status {
+    /// Remote disconnected or remote domain dead
+    Disconnected,
+    /// Connected
+    Connected,
+    /// Server initialized, waiting for client to connect
+    Waiting,
+}
+
+/// A wrapper around a Qubes vchan, which is a stream-oriented, inter-qube
+/// communication channel.  This implementation uses the libvchan C library.
+///
+/// The `Read` implementation of [`Vchan`] does not read from the slice passed
+/// to it, and is safe to call even if that slice is uninitialized memory.
+#[derive(Debug)]
+pub struct Vchan {
+    inner: *mut vchan_sys::libvchan_t,
+}
+
+fn c_int_to_usize(i: c_int) -> usize {
+    assert!(i >= 0, "c_int_to_usize passed negative number");
+    // If u32 doesn’t actually fit in a usize, fail the build
+    let [] = [0; if c_int::MAX as usize as c_int == c_int::MAX {
+        0
+    } else {
+        1
+    }];
+    i as usize
+}
+
+impl Vchan {
+    /// Creates a listening vchan that listens from requests from the given domain
+    /// on the given port.
+    #[inline]
+    pub fn server<T>(
+        domain: T,
+        port: c_int,
+        read_min: usize,
+        write_min: usize,
+    ) -> Result<Self, Error>
+    where
+        u16: From<T>,
+    {
+        Self::server_inner(domain.into(), port, read_min, write_min)
```

Comment:
I assume the separate `server_inner` is to avoid excessive monomorphization. I feel like it's common to put the inner function into the body of the outer function, which clarifies that the scope of the inner function is really this single use.

Also, you don't need to write `where u16: From<T>` – I'd even do `domain: impl Into<u16>`.

(same for `client_inner` below)